lib/bitcoin/lock/LockMonitor.ts

```diff
@@ -0,0 +1,304 @@
+import LockTransactionType from './../enums/LockTransactionType';
+import MongoDbLockTransactionStore from './MongoDbLockTransactionStore';
+import ValueTimeLockModel from './../../common/models/ValueTimeLockModel';
+
+/**
+ * Structure (internal to this class) to track the information about lock.
+ */
+interface LockInformation {
+  currentValueTimeLock: ValueTimeLockModel | undefined;
+  latestSavedLockInfo: LockTransactionModel | undefined;
+
+  // Need to add a 'state'
+}
+
+/**
+ * Encapsulates functionality to monitor and create/remove amount locks on bitcoin.
+ */
+export default class LockMonitor {
+
+  private periodicPollTimeoutId: number | undefined;
+
+  private currentValueTimeLock: LockInformation | undefined;
+
+  private lockResolver: LockResolver;
+
+  constructor (
+    private bitcoinClient: BitcoinClient,
+    private lockTransactionStore: MongoDbLockTransactionStore,
+    private pollPeriodInSeconds: number,
+    private desiredLockAmountInSatoshis: number,
+    private lockPeriodInBlocks: number,
+    private firstLockFeeAmountInSatoshis: number) {
+
+    this.lockResolver = new LockResolver(this.bitcoinClient);
+  }
+
+  /**
+   * Initializes this object by performing the periodic poll tasks.
+   */
+  public async initialize (): Promise<void> {
+    await this.periodicPoll();
+  }
+
+  private async periodicPoll (intervalInSeconds: number = this.pollPeriodInSeconds): Promise<void> {
+
+    try {
+      // Defensive programming to prevent multiple polling loops even if this method is externally called multiple times.
+      if (this.periodicPollTimeoutId) {
+        clearTimeout(this.periodicPollTimeoutId);
+      }
+
+      console.info(`Starting periodic polling for the lock monitor.`);
+      await this.handlePeriodicPolling();
+
+    } catch (e) {
+      const message = `An error occured during periodic poll: ${JSON.stringify(e, Object.getOwnPropertyNames(e))}`;
+      console.error(message);
+    } finally {
+      this.periodicPollTimeoutId = setTimeout(this.periodicPoll.bind(this), 1000 * intervalInSeconds, intervalInSeconds);
+    }
+  }
+
+  private async handlePeriodicPolling (): Promise<void> {
+
+    this.currentValueTimeLock = await this.resolveCurrentValueTimeLock();
+
+    // Note: ALSO need to check the state (pending/confirmed etc) below when the support is added.
+    const validCurrentLockExist = this.currentValueTimeLock.currentValueTimeLock !== undefined;
+
+    const lockRequired = this.desiredLockAmountInSatoshis > 0;
+
+    let resolveCurrentLockAgain = false;
+
+    if (lockRequired && !validCurrentLockExist) {
+      await this.handleCreatingNewLock(this.desiredLockAmountInSatoshis);
+      resolveCurrentLockAgain = true;
+    }
+
+    if (lockRequired && validCurrentLockExist) {
+      await this.handleExistingLockRenewal(
+        this.currentValueTimeLock.currentValueTimeLock!,
+        this.currentValueTimeLock.latestSavedLockInfo!,
+        this.desiredLockAmountInSatoshis);

```

Comment:
Since all the "handle" methods are instance methods, why not let them access the instance properties internally instead of passing them?

Reply:
The 'handle' functions require the 'current-lock-state' inputs and I wanted to limit the usage of the ````currentLockState```` variable to as small piece of code as possible. So now, the 'handle' functions just get the inputs without having to access the ````currentLockState```` variable.